**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Aarón Hernández Jiménez
*   MATRÍCULA: A01642529

---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [1]:
%pip install -q pymysql

Note: you may need to restart the kernel to use updated packages.


`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [2]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [3]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [4]:
query = sqla.text("""
  SELECT *
  FROM movie
  WHERE runtime > 5 * 60   -- runtime está en minutos
""")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,NaN,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


Solo **5 películas** superan las 5 horas (300 min). Casi todas son en realidad *Film Series* o *Limited Series* (colecciones que Netflix registra como un solo título), por eso alcanzan duraciones de hasta 3,892 minutos (~65 horas).

2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [5]:
query = sqla.text("""
  SELECT ROUND(100 * SUM(available_globally = b'0') / COUNT(available_globally), 2) AS porcentaje_solo_eu
  FROM movie
""")
pd.read_sql(query, conn)

,porcentaje_solo_eu
0,83.57


El **83.57 %** de las películas con dato registrado están disponibles únicamente en EU (`available_globally = 0`).

`COUNT(available_globally)` cuenta solo los valores no nulos, por lo que los 672 registros `NULL` quedan fuera del denominador sin necesidad de un `WHERE` adicional; `SUM(condición)` cuenta las filas donde la condición es verdadera (vale 1).

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [6]:
query = sqla.text("""
  SELECT locale, COUNT(*) AS num_peliculas
  FROM movie
  GROUP BY locale
  ORDER BY num_peliculas DESC
""")
pd.read_sql(query, conn)

,locale,num_peliculas
0,NaN,11255
1,en,532


In [7]:
query = sqla.text("""
  SELECT COUNT(*) AS locale_nulos
  FROM movie
  WHERE locale IS NULL
""")
pd.read_sql(query, conn)

,locale_nulos
0,11255


La tabla `movie` solo registra un idioma/región original: **`en`** (inglés, 532 películas). Los otros **11,255 registros tienen `locale` en `NULL`** (≈ 95 % de la tabla), es decir, para la gran mayoría no se conoce el idioma original.

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [8]:
query = sqla.text("""
  SELECT COUNT(*) AS titulos_diferentes
  FROM movie
  WHERE title <> original_title
""")
pd.read_sql(query, conn)

,titulos_diferentes
0,3947


In [9]:
# Desglose de los registros con locale NULL según el estado de original_title
query = sqla.text("""
  SELECT CASE
           WHEN original_title IS NULL THEN 'original_title NULL'
           WHEN title <> original_title THEN 'títulos diferentes'
           ELSE 'títulos iguales'
         END AS caso,
         COUNT(*) AS num_peliculas
  FROM movie
  WHERE locale IS NULL
  GROUP BY caso
  ORDER BY num_peliculas DESC
""")
pd.read_sql(query, conn)

,caso,num_peliculas
0,original_title NULL,7291
1,títulos diferentes,3947
2,títulos iguales,17


In [10]:
query = sqla.text("""
  SELECT title, original_title
  FROM movie
  WHERE locale IS NULL AND title = original_title
""")
pd.read_sql(query, conn)

,title,original_title
0,Recep Ivedik 7,Recep İvedik 7
1,Veronica,Verónica
2,Fantomas,Fantômas
3,Luccas Neto em: O Hotel Magico,Luccas Neto em: O Hotel Mágico
4,Un Dia En El Paraiso,Un día en el paraíso
5,Superlopez,Superlópez
6,Baris Akarsu Merhaba,Barış Akarsu Merhaba
7,Totem,Tótem
8,La Pena Maxima,La pena máxima
9,Mirciulica,Mirciulică


* **3,947** películas tienen `title` distinto de `original_title`, frente a **11,255** con `locale` en `NULL`: **los resultados no coinciden** (faltan 7,308).
* Los registros restantes tienen dos características:
  * **7,291** tienen `original_title` en `NULL`. En SQL, comparar contra `NULL` no da verdadero ni falso sino `NULL`, así que `title <> original_title` las excluye: no hay título original con qué compararlas.
  * **17** tienen ambos títulos "iguales" para MySQL, aunque el original lleva acentos o diacríticos (*Verónica*, *Fantômas*, *Tótem*…). La intercalación (*collation*) de la columna, `utf8mb3_general_ci`, ignora acentos al comparar, por eso los considera iguales. Son títulos en español, turco, portugués, etc.
* Todas las películas con título distinto tienen `locale` en `NULL`, y las 532 en `en` no tienen `original_title`.

**Conclusión:** la suposición **no es válida en general**. Un `NULL` en `locale` indica ausencia de dato, no necesariamente otro idioma: solo en 3,947 de 11,255 casos (≈ 35 %) hay evidencia de un título en otro idioma; en 7,291 no hay información para comprobarlo. Donde sí existe `original_title`, la suposición se cumple (incluso los 17 "iguales" están en otro idioma), pero no puede generalizarse a todos los `NULL`.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [11]:
query = sqla.text("""
  SELECT m.title, v.cumulative_weeks_in_top10
  FROM view_summary v
  JOIN movie m ON m.id = v.movie_id
  WHERE v.cumulative_weeks_in_top10 = (
      SELECT MAX(v2.cumulative_weeks_in_top10)
      FROM view_summary v2
      WHERE v2.movie_id IS NOT NULL
  )
""")
pd.read_sql(query, conn)

,title,cumulative_weeks_in_top10
0,The Boss Baby,22


**The Boss Baby** es la película que más tiempo ha permanecido en el top 10, con **22 semanas acumuladas**. La subconsulta con `MAX()` (en lugar de `ORDER BY … LIMIT 1`) devolvería todas las películas en caso de empate.

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [12]:
query = sqla.text("""
  SELECT title, num_temporadas, posicion
  FROM (
    SELECT t.title,
           COUNT(s.id) AS num_temporadas,
           RANK() OVER (ORDER BY COUNT(s.id) DESC) AS posicion
    FROM tv_show t
    JOIN season s ON s.tv_show_id = t.id
    GROUP BY t.id, t.title
  ) AS ranking
  WHERE posicion <= 5
  ORDER BY posicion, title
""")
pd.read_sql(query, conn)

,title,num_temporadas,posicion
0,Grey's Anatomy,20,1
1,Naruto Shippuden,20,1
2,Heartland (2007),17,3
3,It's Always Sunny in Philadelphia,16,4
4,NCIS,15,5
5,Raw,15,5
6,Supernatural (2005),15,5


Los programas con más temporadas son **Grey's Anatomy** y **Naruto Shippuden** (20 cada uno), **Heartland** (17) e **It's Always Sunny in Philadelphia** (16). El 5.º lugar es un **empate a 15 temporadas** entre *NCIS*, *Raw* y *Supernatural*. Por eso se usó `RANK()`: con `LIMIT 5` el motor elegiría arbitrariamente uno de los tres y ocultaría el empate.

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [13]:
query = sqla.text("""
  SELECT duration, start_date, end_date, COUNT(*) AS num_resumenes
  FROM view_summary
  WHERE duration = 'SEMI_ANNUALLY'
  GROUP BY duration, start_date, end_date
  ORDER BY start_date
""")
pd.read_sql(query, conn)

,duration,start_date,end_date,num_resumenes
0,SEMI_ANNUALLY,2023-07-01,2023-12-31,15763
1,SEMI_ANNUALLY,2024-01-01,2024-06-30,15912


Hay **dos períodos semestrales**:
* **2023-07-01 a 2023-12-31** (segundo semestre de 2023), con 15,763 resúmenes.
* **2024-01-01 a 2024-06-30** (primer semestre de 2024), con 15,912 resúmenes.

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [14]:
query = sqla.text("""
  SELECT s.title, s.season_number, v.views, v.hours_viewed
  FROM view_summary v
  JOIN season s ON s.id = v.season_id
  JOIN tv_show t ON t.id = s.tv_show_id
  WHERE t.title = "Grey's Anatomy"
    AND v.duration = 'SEMI_ANNUALLY'
    AND v.start_date = '2024-01-01'
  ORDER BY v.views DESC, v.hours_viewed DESC
""")
pd.read_sql(query, conn)

,title,season_number,views,hours_viewed
0,Grey's Anatomy: Season 1,1,3600000,23400000
1,Grey's Anatomy: Season 2,2,3100000,60600000
2,Grey's Anatomy: Season 3,3,2900000,53700000
3,Grey's Anatomy: Season 5,5,2900000,49300000
4,Grey's Anatomy: Season 4,4,2900000,35300000
5,Grey's Anatomy: Season 6,6,2800000,48000000
6,Grey's Anatomy: Season 7,7,2700000,43000000
7,Grey's Anatomy: Season 8,8,2600000,44700000
8,Grey's Anatomy: Season 9,9,2500000,43300000
9,Grey's Anatomy: Season 10,10,2400000,41700000


**Interpretación:**
* Las vistas **disminuyen casi de forma monótona con el número de temporada**: la temporada 1 lidera con 3.6 M y la 18 cae a 1.7 M. Es el patrón típico de abandono en series largas: muchos usuarios empiezan desde el principio y en cada temporada se van quedando menos.
* `views` se calcula como horas vistas ÷ duración de la temporada. Por eso la temporada 1, que es corta (387 min), tiene más *vistas* que la 2 aunque esta acumula muchas más *horas* (60.6 M vs 23.4 M).
* Las temporadas 3, 4 y 5 empatan en 2.9 M de vistas; el desempate por `hours_viewed` las ordena 3 → 5 → 4.
* La **temporada 19** (2.0 M) rompe la tendencia y supera a las 15–18, probablemente por ser la más reciente completa en el catálogo.
* La **temporada 20** apenas registra 0.1 M probablemente porque se estrenó durante ese semestre (y en el registro tiene 431 min, menos de la mitad que las demás): fue visible muy poco tiempo.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [15]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [16]:
# 1. Unir las tablas siguiendo las llaves foráneas: view_summary.season_id → season.id → tv_show.id
merged_df = (
    view_summary
    .merge(season, left_on='season_id', right_on='id', suffixes=('', '_season'))
    .merge(tv_show, left_on='tv_show_id', right_on='id', suffixes=('', '_tv_show'))
)

# 2. Filtrar por programa (title_tv_show, ya que title es el de la temporada), período y fechas del primer semestre de 2024
merged_df['start_date'] = pd.to_datetime(merged_df['start_date'])

mask = (
    (merged_df['title_tv_show'] == "Grey's Anatomy") &
    (merged_df['duration'] == 'SEMI_ANNUALLY') &
    (merged_df['start_date'] >= '2024-01-01') &
    (merged_df['start_date'] <= '2024-06-30')
)

# 3. Ordenar por views y hours_viewed (descendente) y seleccionar columnas requeridas
resultado = (
    merged_df[mask]
    .sort_values(by=['views', 'hours_viewed'], ascending=[False, False])
    [['title', 'season_number', 'views', 'hours_viewed']]
    .astype({'season_number': int})
    .reset_index(drop=True)
)
resultado

,title,season_number,views,hours_viewed
0,Grey's Anatomy: Season 1,1,3600000,23400000
1,Grey's Anatomy: Season 2,2,3100000,60600000
2,Grey's Anatomy: Season 3,3,2900000,53700000
3,Grey's Anatomy: Season 5,5,2900000,49300000
4,Grey's Anatomy: Season 4,4,2900000,35300000
5,Grey's Anatomy: Season 6,6,2800000,48000000
6,Grey's Anatomy: Season 7,7,2700000,43000000
7,Grey's Anatomy: Season 8,8,2600000,44700000
8,Grey's Anatomy: Season 9,9,2500000,43300000
9,Grey's Anatomy: Season 10,10,2400000,41700000


El resultado con Pandas coincide con el de la consulta SQL de la pregunta 8.

`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [17]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

El archivo `consultancy.json` debe estar en la misma carpeta que esta libreta (se lee con ruta relativa). No se publica en el repositorio porque contiene una clave privada.

In [18]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [19]:
docs = list(db.collection('EMPLOYEE').stream())
employee = pd.DataFrame([d.to_dict() for d in docs], index=pd.Index([d.id for d in docs], name='id'))
employee[['emp_fname', 'emp_lname', 'emp_hiredate']]

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [20]:
firebase_admin.delete_app(firebase_admin.get_app())

In [21]:
conn.close()

---

**Declaración de uso de IA**

*   Anthropic. (2026). *Claude (Opus 5.5)* [Modelo de lenguaje grande], utilizado para generación de consultas SQL, depuración y redacción de interpretaciones. https://claude.ai/

---